frame/assistants/shadow-liketh/src/shadow.rs

```diff
@@ -0,0 +1,189 @@
+        let mmr_root = self.mmr_root(block_number).await?;
+        Ok(HeaderParcel { mmr_root, header })
+    }
+
+    pub async fn receipt(
+        &self,
+        tx: impl AsRef<str>,
+        last: u64,
+    ) -> ShadowComponentReuslt<EthereumReceiptWithMMRProof> {
+        let resp = self
+            .http
+            .get(&format!(
+                "{}/ethereum/receipt/{}/{}",
+                &self.config.endpoint,
+                tx.as_ref(),
+                last
+            ))
+            .send()
+            .await?;
+        if resp.status() == reqwest::StatusCode::INTERNAL_SERVER_ERROR {
+            return Err(ShadowComponentError::InternalServer(resp.text().await?));
+        }
+        let result: serde_json::Value = resp.json().await?;
+        if let Some(err) = result.get("error") {
+            let msg = err.as_str().ok_or_else(|| {
+                ShadowComponentError::Cusom("Failed parse error message".to_string())
+            })?;
+            return Err(ShadowComponentError::Cusom(msg.to_owned()));
+        }
+        let receipt: EthereumReceiptJson = serde_json::from_value(result)?;
+        let header = &receipt.header;
+
+        let (member_leaf_index, last_leaf_index) = (header.number, last - 1);
+        let proof = self.mmr_proof(member_leaf_index, last_leaf_index).await?;
+        let mmr_proof = MMRProofJson {
+            member_leaf_index,
+            last_leaf_index,
+            proof,
+        };
+        Ok(EthereumReceiptWithMMRProof { receipt, mmr_proof })
+    }
+}
+
+impl Shadow {
+    fn calc_peaks(
+        &self,
+        positions: Vec<MMRPosition>,
+    ) -> ShadowComponentReuslt<Vec<(u64, [u8; 32])>> {
+        positions
+            .iter()
+            .map(|item| {
+                let array: arrayvec::ArrayVec<_, 32> = item.hash.clone().into_iter().collect();
+                match array
+                    .into_inner()
+                    .map_err(|e| ShadowComponentError::MMR(format!("{:?}", e)))
+                {
+                    Ok(v) => Ok((item.id, v)),
+                    Err(e) => Err(e),
+                }
+            })
+            .collect::<Vec<ShadowComponentReuslt<(u64, [u8; 32])>>>()
+            .into_iter()
+            .collect::<ShadowComponentReuslt<Vec<(u64, [u8; 32])>>>()
+    }
+
+    async fn query_position(&self, positions: Vec<u64>) -> ShadowComponentReuslt<Vec<MMRPosition>> {
+        let query = self.read_graphql("mmr_position.query.graphql")?;
+
+        let vars = QueryPositionVars { positions };
+        let response = self
+            .gql
+            .query_with_vars_unwrap::<TheGraphResponse, QueryPositionVars>(query, vars)
+            .await
+            .map_err(ShadowComponentError::from)?;
+
+        match response {
+            TheGraphResponse::NodeEntities(data) => Ok(data),
+        }
+    }
+
+    //
+    pub async fn mmr_root(&self, leaf_index: u64) -> ShadowComponentReuslt<[u8; 32]> {
```

Comment:
Can directly retrieve from on chain storage using optional parameter `at_block`

Reply:
better to directly query from block header, or return together with proof.

Ref: 
https://github.com/darwinia-network/darwinia-common/blob/17ce2d0d9668124bd707be92ff62c222aff99b25/frame/header-mmr/src/lib.rs#L164

Reply:
```
    /// get mmr root
    pub async fn get_mmr_root(&self, leaf_index: u32) -> ClientResult<subxt::sp_core::H256> {
        let block_number = leaf_index + 1;
        let header = self.header_by_number(block_number).await?;

        let mmr_root = if let Some(header) = header {
            // get digest_item from header
            let log = header
                .digest()
                .logs()
                .iter()
                .find(|&x| x.as_other().is_some());
            if let Some(digest_item) = log {
                // get mmr_root from log
                let parent_mmr_root = digest_item.as_other().unwrap().to_vec();
                let parent_mmr_root = &parent_mmr_root[4..];
                if parent_mmr_root.len() != 32 {
                    return Err(ClientError::WrongMmrRootInDarwiniaHeader(
                        array_bytes::bytes2hex("", &parent_mmr_root),
                        block_number,
                    ));
                }
                let mut mmr_root: [u8; 32] = [0; 32];
                mmr_root.copy_from_slice(parent_mmr_root);
                subxt::sp_core::H256(mmr_root)
            } else {
                return Err(ClientError::NoMmrRootInDarwiniaHeader(block_number));
            }
        } else {
            return Err(ClientError::FailedToFetchDarwiniaHeader(block_number));
        };
        Ok(mmr_root)
    }
```